---
title: "SharedRMarkdown"
output: html_document
date: "2024-11-27"
---

Outline:

This R Markdown script provides the code needed to reproduce the RNA-sequencing analyses conducted in this study. The script below can be utilised to replicate the data and analysis used in this work.

# Initial setup of the Work environment and Data:

1. Initially, we logged into the CREATE note and navigated to our scratch place


```{bash}

srun -p msc_appbio --pty /bin/bash

/scratch_tmp/grp/msc_appbio/group4_tmp/
```
2. Preparing the workspace 
Create new directories to store files within project 

```{bash}
mkdir raw_data_rna
mkdir processed_data 
mkdir ref_genome

```
# Downloading fastq files with the command line
The wget command can be used to transfer data from BioStudies (https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-9499/sdrf) and from ENA (https://www.ebi.ac.uk/ena/browser/view/PRJEB41109?show=reads)

1. Create a nano script containing all the URLs for the fastq files we need 

```{bash}
nano fastqLink.txt # for RNA-seq 
nano bam_urls.txt # for Genome-seq 

```

2. Copying the data that is used into the nano scripts

```{bash}
wget -c -i bam_urls.txt # for Genome-seq
wget -c -i fastqLink.txt # for RNA-seq
```
'-i:' This option ensures that the URLs can be read from the files bam_urls.txt and fastqLink.txt. 

The ouput should include a list of bam files for the Genome-seq data and a list of fastq.gz files for the RNA-seq data. 

# Installing the tools needed for pur environment using Conda

1. Create nano script to create a Conda environment with specific versions - for both genome seq and RNA-seq

```{bash}
nano environment.sh

# In the nano script called environment.sh

#!/bin/bash

echo "start of the pipeline"

# Add conda channels
conda config --add channels bioconda
conda config --add channels conda-forge

# Create a new conda environment with specific versions
conda create -n bioenv cutadapt=2.3 faqcs=2.08 star= 2.5.2a fdrtool= 1.2.15 -y

#Load Fastqc

module load fastqc

# Print a message about how to activate the environment
echo "To activate your environment, run: conda activate bioenv" 


echo "end of the pipeline"

Then we need to activate it using 
./environment.sh
But permission is denied
conda activate bioenv


```


# Quality Control of our files, using fastqc

1. Create a Nano script to fastq the fasta files for RNA-Seq

```{bash}
nano fastqc_RNA

# In the nano script

#!/bin/bash

echo "start of the pipeline"

#Load fastqc - if not loaded previously 
module load fastqc 

# Define base and results directories

baseDirectory="/scratch_tmp/grp/msc_appbio/group4_tmp/raw_data_rna/fastq"
resultsDirectory="/scratch/grp/msc_appbio/group4_tmp/processed_data/"

# Create the results directory 
mkdir -p "$resultsDirectory"

# Run Fastqc on all .fastq.gz files in baseDirectory
fastqc -o "$resultsDirectory" -t 4 "$baseDirectory"/*.fastq.gz 
#(-o represents the output directory)

echo "end of the pipeline"

```

If the fastqc.html files look good i.e. have good quality reads and do not require adapter trimming, you can proceed with the raw .fastq files without further processing. 

#Alignment with STAR

1. Obtaining the reference genome

```{bash}
wget -c -i https://ftp.ensemblgenomes.ebi.ac.uk/pub/fungi/release-60/fasta/fungi_ascomycota1_collection/saccharomyces_cerevisiae_cen_pk113_7d_gca_000269885/dna/Saccharomyces_cerevisiae_cen_pk113_7d_gca_000269885.ASM26988v1.dna.toplevel.fa.gz #should be directed to the ref_genome directory

gunzip ref_genome  #gunzips the reference genome 
```




For git:
ssh-keygen -t ed25519 -C "eraaziz13@gmail.com"

cat /users/k24088153/.ssh/id_ed25519.pub

git config --global --add safe.directory /scratch_tmp/grp/msc_appbio/group4_tmp/MSc_appbio_group4
git status 
This was so that we could all access our git repository. 

When we change a script that is on the repository we do:

git add with file name 
git commit with file name
git push 

